test/functional/wallet_upgradewallet.py

```diff
@@ -0,0 +1,125 @@
+            releases_path + "/v0.15.2/bin/bitcoind",
+        ], binary_cli=[
+            self.options.bitcoincli,
+            releases_path + "/v0.16.3/bin/bitcoin-cli",
+            releases_path + "/v0.15.2/bin/bitcoin-cli",
+        ])
+        # adapt bitcoin.conf, because older bitcoind's don't recognize config sections
+        adjust_bitcoin_conf_for_pre_17(self.nodes[1].bitcoinconf)
+        adjust_bitcoin_conf_for_pre_17(self.nodes[2].bitcoinconf)
+        self.start_nodes()
+
+    def run_test(self):
+        self.nodes[0].generatetoaddress(101, self.nodes[0].getnewaddress())
+        # # Sanity check the test framework:
+        res = self.nodes[0].getblockchaininfo()
+        assert_equal(res['blocks'], 101)
+        node_master = self.nodes[0]
+        v16_3_node  = self.nodes[1]
+        v15_2_node  = self.nodes[2]
+
+        self.log.info("Test upgradewallet RPC...")
+        # Prepare for copying of the older wallet
+        node_master_wallets_dir = os.path.join(node_master.datadir, "regtest/wallets")
+        v16_3_wallets_dir       = os.path.join(v16_3_node.datadir, "regtest/wallets")
+        v15_2_wallets_dir       = os.path.join(v15_2_node.datadir, "regtest")
+        # stop all nodes
+        self.stop_node(0)
+        self.stop_node(1)
+        self.stop_node(2)
+
+        # Copy the 0.16.3 wallet to the last Bitcoin Core version and open it:
+        shutil.rmtree(node_master_wallets_dir)
+        shutil.copytree(
+            v16_3_wallets_dir,
+            node_master_wallets_dir
+        )
+        self.restart_node(0, ['-nowallet'])
+
+        node_master.loadwallet(node_master_wallets_dir)
+        wallet_name = node_master.getwalletinfo()["walletname"]
+        wallet = node_master.get_wallet_rpc(wallet_name)
+        old_version = wallet.getwalletinfo()["walletversion"]
+        # calling upgradewallet without version arguments
+        # should return nothing if successful
+        assert_equal(wallet.upgradewallet(), "")
+        new_version = wallet.getwalletinfo()["walletversion"]
+        # upgraded wallet version should be greater than older one
+        assert_greater_than(new_version, old_version)
+
+        self.stop_node(0)
+
+        # Copy the 0.15.2 wallet to the last Bitcoin Core version and open it:
+        shutil.rmtree(node_master_wallets_dir)
+        shutil.copytree(
+            v15_2_wallets_dir,
+            node_master_wallets_dir
+        )
+        self.restart_node(0, ['-nowallet'])
+
+        node_master.loadwallet(node_master_wallets_dir)
+        wallet_name = node_master.getwalletinfo()["walletname"]
+        wallet = node_master.get_wallet_rpc(wallet_name)
+        # calling upgradewallet with explicit version number
+        # should return nothing if successful
+        assert_equal(wallet.upgradewallet(169900), "")
+        new_version = wallet.getwalletinfo()["walletversion"]
+        # upgraded wallet should have version 169900
+        assert_equal(new_version, 169900)
```

Comment:
Is this now an HD wallet? If yes, would be nice to check that the master key hash is present in the wallet info (or something like that)

Reply:
Have added appropriate tests.